lib/kazan/watcher.ex

```
@@ -0,0 +1,164 @@
+  def start_link(%Kazan.Request{} = request, opts) do
+    send_to = Keyword.fetch!(opts, :send_to)
+    rv = Keyword.fetch!(opts, :resource_version)
+    GenServer.start_link(__MODULE__, [request, rv, send_to])
+  end
+
+  @impl GenServer
+  def init([request, rv, send_to]) do
+    Logger.info "Watcher init pid: #{inspect self()} rv: #{rv} request: #{inspect request}"
+    {:ok, id} = Kazan.Client.run(request, stream_to: self())
+    {:ok, %State{id: id, request: request, rv: rv, send_to: send_to}}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncStatus{code: 200}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncHeaders{}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncChunk{chunk: chunk, id: request_id},
+                  %State{id: id, buffer: buffer, rv: rv, send_to: send_to} = state) do
+    if request_id != id do
+      Logger.warn "Received chunk for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
+    else
+
+      {lines, buffer} =
+        buffer
+        |> LineBuffer.add_chunk(chunk)
+        |> LineBuffer.get_lines()
+
+        new_rv = process_lines(lines, rv, send_to)
+
+      {:noreply, %State{state | buffer: buffer, rv: new_rv}}
+    end
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.Error{reason: {:closed, :timeout}}, state) do
+    Logger.info "Received Timeout with id: #{inspect self()}"
+    {:noreply, restart_connection(state)}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncEnd{id: request_id},
+                  %State{id: id} = state) do
+    Logger.info "Received AsyncEnd for id: #{inspect request_id}"
+    if request_id != id do
+      Logger.warn "Received AsyncEnd message for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
```

Again, this shouldn't really happen, so I'm wondering if the correct thing to do would be to let it crash so supervisor can restart things in a hopefully-good state.

---

Yes